### PR TITLE
Assignment4/submission

### PR DESCRIPTION
**Pull request by AntonScheitler**

### CODE REVIEW

File: src/computations/temperatures/TemperatureComputations.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * @brief a class wrapping position-computing functions and their utility functions
+ */
+class TemperatureComputations {
+public:
+    static void updateTemp(ParticleContainer &particles, double targetTemp, double maxDeltaTemp);
+
+    static void initTemp(ParticleContainer &particles, double averageVelocity, double initialTemp);
+
+    static double calculateCurrentSystemTemp(ParticleContainer &particles);
```
Comment (SamNewcome) [inline src/computations/temperatures/TemperatureComputations.h:14]: Missing documentation

File: src/computations/temperatures/TemperatureComputations.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include "particle/container/ParticleContainer.h"
+
+/**
+ * @brief a class wrapping position-computing functions and their utility functions
+ */
```
Comment (SamNewcome) [inline src/computations/temperatures/TemperatureComputations.h:7]: This documentation is wrong

File: src/computations/temperatures/TemperatureComputations.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#include "utils/ArrayUtils.h"
+#include "utils/MaxwellBoltzmannDistribution.h"
+#include "spdlogConfig.h"
+
+
+void TemperatureComputations::initTemp(ParticleContainer& particles, double averageVelocity, double initialTemp) {
+    for (auto it = particles.begin(); *it != *(particles.end()); it->operator++()) {
+        Particle& particle = **it;
+        double factor = sqrt(initialTemp / particle.getM());
+        std::array<double, 3> maxBoltzVelocity = maxwellBoltzmannDistributedVelocity(averageVelocity, 3);
```
Comment (SamNewcome) [inline src/computations/temperatures/TemperatureComputations.cpp:11]: This is still wrong from WS2! The second argument of `maxwellBoltzmannDistributedVelocity` is the dimension which should be set to `2` for 2D simulations. As WS5 is 3D, keep this as 3D (but then remove the forcing the z-velocity to 0)

File: src/computations/temperatures/TemperatureComputations.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#include "./TemperatureComputations.h"
+#include "utils/ArrayUtils.h"
+#include "utils/MaxwellBoltzmannDistribution.h"
+#include "spdlogConfig.h"
+
+
+void TemperatureComputations::initTemp(ParticleContainer& particles, double averageVelocity, double initialTemp) {
```
Comment (SamNewcome) [inline src/computations/temperatures/TemperatureComputations.cpp:7]: You should not be using the old Brownian motion average velocity scaled by the new temperature-based factor - the new temperature-based factor replaces the old Brownian motion average velocity.

The consequence of this is that none of your WS4 simulations uses any Brownian motion because you use an old Brownian motion average of the velocity of 0, which gets scaled to 0 regardless of the initial temperature.

This means that you will start your simulations at a very cold temperature, which then gets scaled at the end of the first iteration by the thermostat to the expected initial temperature. As this scaling is larger than it should be, you get this very "dramatic" start to your simulations, which may explain why your falling drop kinda "explodes" outwards.

Instead, remove the old brownian motion average velocity and provide `factor` to `maxwellBoltzmannDistributedVelocity` in place of `averageVelocity`

File: test/computations/temperatures/TemperatureComputationsTest.cpp
Comment (SamNewcome) [inline test/computations/temperatures/TemperatureComputationsTest.cpp:1]: Lots of wrong documentation/test descriptions

Comment (SamNewcome) [reply test/computations/temperatures/TemperatureComputationsTest.cpp:1]: No unit test for new Brownian motion mechanism (might have spotted the mistake you made if you had this)

File: src/particle/container/ParticleContainerLinkedCell.h [hunk 5]
```diff
@@ -70,13 +70,36 @@ class ParticleContainerLinkedCell : public ParticleContainer {
...
+     * @param point2 the second point
+     * @param v the non-periodic vector that points from point1 to point2
+     */
+    void getPeriodicDistanceVector(const std::array<double, 3> &point1, const std::array<double, 3> &point2,
+                                   std::array<double, 3> &v);
+
+    /**
+     * @brief applies periodic boundaries to the specified array
+     * @param coord the coordinate array to apply the periodic boundaries to 
+     */
```
Comment (SamNewcome) [inline src/particle/container/ParticleContainerLinkedCell.h]: This is too vague. "Applying periodic boundaries" could mean anything from looping coordinates outside the domain to their periodic copy inside the domain (which is what this does) or handling the entire mechanic of periodic boundaries.

File: src/particle/container/ParticleContainerLinkedCell.h [hunk 6]
```diff
@@ -70,13 +70,36 @@ class ParticleContainerLinkedCell : public ParticleContainer {
...
 
+    /**
+     * @brief changes the supplied vector to a periodic one, if the supplied points are on opposite sides of a periodic boundary. If the points aren't on opposite sides, 
+     * or the boundary is non-periodic, the supplied vector remains unchanged
+     * @param point1 the first point
+     * @param point2 the second point
+     * @param v the non-periodic vector that points from point1 to point2
+     */
+    void getPeriodicDistanceVector(const std::array<double, 3> &point1, const std::array<double, 3> &point2,
+                                   std::array<double, 3> &v);
```
Comment (SamNewcome) [inline src/particle/container/ParticleContainerLinkedCell.h]: `v` is not a verbose enough name.

File: test/particle/ParticleContainerLinkedCellTest.cpp
Comment (SamNewcome) [inline test/particle/ParticleContainerLinkedCellTest.cpp:1]: I would suggest some corner case tests for both periodic/periodic and periodic/reflecting corners.

I don't think this is as important for your code as other codes because your implementation requires no special handling of corners, but it is still good.

File: src/computations/forces/ForceComputations.h [hunk 7]
```diff
@@ -1,44 +1,55 @@
...
+                                                   double cutoff);
+
+
+    /**
+     * @brief updates the old force and resets the current force for all particles
+     * @param particles the particles to reset the forces of
+     */
+    static void resetForces(ParticleContainer &particles);
+
+    static void addExternalForces(ParticleContainer &particles, double grav);
```
Comment (SamNewcome) [inline src/computations/forces/ForceComputations.h:42]: No documentation :(

Comment (SamNewcome) [reply src/computations/forces/ForceComputations.h:42]: Also no unit test

File: src/computations/forces/ForceComputations.cpp [hunk 8]
```diff
@@ -47,12 +62,15 @@ void ForceComputations::computeLennardJonesPotential(ParticleContainer &particle
     }
 }
 
-void ForceComputations::computeLennardJonesPotentialCutoff(ParticleContainer &particles, double epsilon, double sigma, double cutoff) {
+void ForceComputations::computeLennardJonesPotentialCutoff(ParticleContainerLinkedCell &particles, double epsilon,
+                                                           double sigma, double cutoff) {
```
Comment (SamNewcome) [inline src/computations/forces/ForceComputations.cpp:66]: This doesn't apply to the mixing rule. Instead the non-cutoff variant applies the mixing rule (which still implements the cutoff?). The non-cutoff variant however does not apply periodic or reflecting boundaries nor is it used with the thermostat. 

I am clearly misunderstanding something, because otherwise you could not achieve the RT videos. Can you explain what I am not understanding?

Comment (SamNewcome) [reply src/computations/forces/ForceComputations.cpp:66]: Also, I don't see a unit test for mixing.

File: src/io/inputReader/XMLReader/XMLFileReader.cpp [hunk 9]
```diff
@@ -40,28 +42,49 @@ namespace inputReader {
...
                 simData.setParticles(std::move(containerDirectSum));
             }
             ParameterParser::readParams(simData, simParser);
+            ParameterParser::readThermo(simData, simParser);
+
+            if (simParser->parameters()->import_checkpoint().present()) {
+                CheckpointReader checkpointReader(simData);
+                type = checkpointReader.readCheckpointFile(simData,
+                                                           simParser->parameters()->import_checkpoint()->file_path().c_str());
+            }
```
Comment (SamNewcome) [inline src/io/inputReader/XMLReader/XMLFileReader.cpp:85]: No unit test for this

File: src/particle/cell/Cell.cpp [hunk 10]
```diff
@@ -11,26 +11,32 @@ Cell::Cell(bool isBoundaryArg, size_t cellIdArg) {
...
 
-void Cell::addParticle(const Particle &particle) {
-    particles.push_back(particle);
+void Cell::addParticleIdx(const size_t idx) {
+    //auto it = std::find_if(particleIndices.begin(), particleIndices.end(),
+    //                       [idx](const size_t currIdx) { return currIdx == idx; });
+    //if (it != particleIndices.end()) {
+    //    SPDLOG_ERROR("SOMETHING TERRIBLE HAS HAPPENED :(");
+    //    exit(EXIT_FAILURE);
+    //}
```
Comment (SamNewcome) [inline src/particle/cell/Cell.cpp]: deadcode

**Review by SamNewcome — APPROVED** [review]
Well done!

It particular, I like your approach to the boundary conditions. It's much simpler than most groups which makes it easier to read, understand, and maintain. Particularly, you do not need to handle edges/corners specially and your code will easily migrate to 3D.

The main issue with the code is a lot of missing unit tests and documentation. Remember, all code should be documented and every new feature should have a unit test. Also, there's some issues with the Brownian motion.

A minor issue which I can't comment on within the code is that you have two XML folders in io with no obvious reason for the distinction from folder names. Either merge the folders or give them names which make it clearer what the distinction is.

There are however some things which aren't clear to me and I have some questions:

- I do not understand how you got the Rayleigh-Taylor simulations. You have only implemented mixing in the non-cutoff/LC variant of the simulation, which doesn't have features like a thermostat or BCs. Am I misunderstanding this? Please explain.
- Due to insufficient instructions, I cannot check the checkpointing mechanism. Please provide instructions for how I can run an equilibration simulation, save a checkpoint, and then run the falling drop simulation.
- Due to insufficient instructions, I cannot compile on CoolMUC4. The report mentions an earlier version of the code was used but no commit ID was provided. Please provide the commit ID used and any build instructions including any modules that need to be loaded. This also holds for `icpx` - if you could not compile with this let me know.

File: src/computations/temperatures/TemperatureComputations.cpp (hunk 3 above)
Comment (FabianHaeusel) [reply src/computations/temperatures/TemperatureComputations.cpp:11]: Yes, I think we may have overlooked that you can pass the number of dimensions as a parameter.

File: src/particle/container/ParticleContainerLinkedCell.h (hunk 5 above)
Comment (AntonScheitler) [reply src/particle/container/ParticleContainerLinkedCell.h]: done

File: src/particle/container/ParticleContainerLinkedCell.h (hunk 6 above)
Comment (AntonScheitler) [reply src/particle/container/ParticleContainerLinkedCell.h]: done

File: src/computations/forces/ForceComputations.cpp (hunk 8 above)
Comment (AntonScheitler) [reply src/computations/forces/ForceComputations.cpp:66]: I see why this is confusing. This is something we also realized after submission. We worked in parallel on implementing the mixing rule and periodic boundaries and when merging the branches we accidentally ended up with two functions. One had the mixing rule and the other had the periodic boundaries. We overlooked that the new features were in different functions and created the video using the cutoff variant

**Comment by FabianHaeusel** [comment]
> * Due to insufficient instructions, I cannot compile on CoolMUC4. The report mentions an earlier version of the code was used but no commit ID was provided. Please provide the commit ID used and any build instructions including any modules that need to be loaded. This also holds for `icpx` - if you could not compile with this let me know.

The commit id of the version of the code which we ran on CoolMUC4 is f2ef8a2fd0f6403be52563b11eb3e11b2601f31a. (The last commit before the code cleanup). This is also the version for which we added the benchmark results in the report. But we failed to mention the commit id.
Build instructions:
(in directory MolSim)
```
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=RELEASE
make
```
Afaik we didn't load any further modules.

**Comment by SamNewcome** [comment]
Okay, this fixes it. Please don't forget to send instructions for the checkpointing.

File: src/computations/temperatures/TemperatureComputations.cpp (hunk 4 above)
Comment (FabianHaeusel) [reply src/computations/temperatures/TemperatureComputations.cpp:7]: Thanks for the correction, we were confused by the term factor so we thought it would be a factor like in the velocity scaling. Now with the corrected thermostat we still have the effect that the falling drop explodes outwards, so this was probably not the origin of that and we will need to look further into it (the fix is being done with worksheet 5, so not on this branch).

File: test/computations/temperatures/TemperatureComputationsTest.cpp
Comment (FabianHaeusel) [reply test/computations/temperatures/TemperatureComputationsTest.cpp:1]: Yes. This would have probably been very useful. We added that.

File: src/computations/forces/ForceComputations.h (hunk 7 above)
Comment (FabianHaeusel) [reply src/computations/forces/ForceComputations.h:42]: done
